src/scores/processing/block_bootstrap.py

```diff
@@ -0,0 +1,410 @@
+"""
+Functions for performing block bootstrapping of arrays. This is inspired from
+https://github.com/dougiesquire/xbootstrap with modifications to make the functions more
+testable and also consistent with the scores package.
+"""
+import math
+from collections import OrderedDict
+from itertools import chain, cycle, islice
+from typing import Dict, List, Tuple, Union
+
+import numpy as np
+import xarray as xr
+
+from scores.typing import XarrayLike
+
+MAX_CHUNK_SIZE_MB = 200
+
+
+def _get_blocked_random_indices(
```

Comment:
We may need to walk through this logic. Currently the logic feels a bit coupled (to me anyway).

For example, it seems that `shape` is mutable so it's pretty tricky for me to evaluate what it's trying to do. `shape[prev_ax] = math.ceil(shape[prev_ax] / b)` implies that `/ b` could happen multiple times depending on the number of axes. Doing it once implies `shape / block_size = num_blocks`, but doing it again, `block_size / num_blocks` implies there's some iterative reduction in the sampling space for that axis. But this reduction is controlled by a iterative call in another function: `_n_nested_blocked_random_indices`.

There may be more explicit ways for you to rearrange the logic in this function for clarity? Maybe a tree structure can help with organizing the data? Example:

```                                                    
     axis_1 │       sample_1        sample_2    ...  N_1   
            │          │               │                   
            │     ┌────┼────┐      ┌───┼───┐               
            │     │    │    │      │   │   │               
            │     │    │    │      │   │   │               
     axis_2 │   s_1   s_2   s_n    ▼   ▼   ▼    ...  N_2   
            │    │     │     │     │   │   │               
       ...  │    │     │     │     │   │   │               
            │  ┌─┼─┐ ┌─┼─┐   │   ┌─┼─┬─┼─┬─┼─┐             
            │  │ │ │ │ │ │   │   │ │ │ │ │ │ │             
     axis_n │  ▼ ▼ ▼ ▼ ▼ ▼   ▼   ▼ ▼ ▼ ▼ ▼ ▼ ▼  ...  N_n   
   ─────────┴───────────────────────────────────────────   
              samples--->                                                                  
                                                                                                   
    - each leaf contains n-tuple representing the combined sample index pointing to a single entry.
    - repeat * n_iter and stack if we want multiple of these.
    - each axis level is randomly sampled based on its block size, for each parent sample/block.
```

But this also may not be right if there are restrictions on what the underlying algorithm should extract and how it needs to do it.

Happy to discuss. Maybe once @tennlee also has had a look...